src/ILLink.Shared/XmlAttributeParser.cs

```diff
@@ -0,0 +1,317 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Xml;
+using System.Xml.Linq;
+using System.Xml.XPath;
+
+namespace ILLink.Shared
+{
+	public static class LinkAttributes
+	{
+		public static List<ITopLevelNode> ProcessXml (XDocument doc)
+		{
+			try {
+				XPathNavigator nav = doc.CreateNavigator ();
+				return NodeBase.ProcessRootNodes (nav);
+			}
+			// TODO: handle the correct exceptions correctly
+			catch (ArgumentException) {
+				// XML doesn't have a <linker> tag and is invalid. The document should have been validated before 
+				return new List<ITopLevelNode> ();
+			}
+		}
+
+		public record AttributeNode : NodeBase
+		{
+			public string FullName;
+			public string Internal;
+			public string Assembly;
+			public List<IAttributeArgumentNode> Arguments;
+			public List<AttributePropertyNode> Properties;
+			public List<AttributeFieldNode> Fields;
+
+			public AttributeNode (XPathNavigator nav) : base (nav)
+			{
+				var arguments = new List<IAttributeArgumentNode> ();
+				foreach (XPathNavigator argNav in nav.SelectChildren (ArgumentElementName, "")) {
+					if (ProcessArgument (argNav) is AttributeArgumentNode arg) {
+						arguments.Add (arg);
+					}
+				}
+				var properties = new List<AttributePropertyNode> ();
+				foreach (XPathNavigator propertyNav in nav.SelectChildren (ArgumentElementName, "")) {
+					var prop = new AttributePropertyNode (name: GetName (propertyNav), value: propertyNav.Value);
+					properties.Add (prop);
+				}
+				var fields = new List<AttributeFieldNode> ();
+				foreach (XPathNavigator fieldNav in nav.SelectChildren (ArgumentElementName, "")) {
+					var field = new AttributeFieldNode (name: GetName (fieldNav), value: fieldNav.Value);
+					fields.Add (field);
+				}
+				FullName = GetFullName (nav);
+				Internal = GetAttribute (nav, "internal");
+				Assembly = GetAttribute (nav, AssemblyElementName);
+				Arguments = arguments;
+				Properties = properties;
+				Fields = fields;
+
+				static IAttributeArgumentNode ProcessArgument (XPathNavigator argNav)
+				{
+					var type = argNav.GetAttribute ("type", "");
+					if (type == "") type = "System.String";
+					if (type == "System.Object") {
+						foreach (XPathNavigator innerArgNav in argNav.SelectChildren (ArgumentElementName, "")) {
+							if (ProcessArgument (innerArgNav) is AttributeArgumentNode innerArg) {
+								return new AttributeArgumentBoxNode (innerArg);
+							} else {
+								// error
+								throw new ArgumentException ("Arguments to attribute cannot be wrapped in more than one type");
```

Comment:
See my other comment about `ArgumentException` - same question here. Will the schema validation catch this? We should add tests for this to ensure it behaves as expected.

Reply:
I updated the parsing to record diagnostics in the nodes of the tree built in ParseXml, then the analyzer emits as it traverses the tree. This way, I can remove the exceptions from ParseXml.